token/core/common/validator_auditing.go

```diff
@@ -0,0 +1,48 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package common
+
+import (
+	"github.com/hyperledger-labs/fabric-token-sdk/token/driver"
+	"github.com/pkg/errors"
+)
+
+func AuditingSignaturesValidate[P driver.PublicParameters, T any, TA driver.TransferAction, IA driver.IssueAction, DS driver.Deserializer](ctx *Context[P, T, TA, IA, DS]) error {
+	if len(ctx.PP.Auditors()) == 0 {
+		// enforce no auditor signatures are attached
+		if len(ctx.TokenRequest.AuditorSignatures) != 0 {
+			return errors.New("auditor signatures are not empty")
+		}
+		return nil
+	}
+
+	auditors := ctx.PP.Auditors()
+	for _, auditorSignature := range ctx.TokenRequest.AuditorSignatures {
+		auditor := auditorSignature.Identity
+		// check that issuer of this issue action is authorized
+		found := false
+		for _, target := range auditors {
+			if auditor.Equal(target) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return errors.Errorf("auditor [%s] is not in auditors", auditor)
+		}
+
+		verifier, err := ctx.Deserializer.GetAuditorVerifier(auditor)
+		if err != nil {
+			return errors.Wrapf(err, "failed to deserialize auditor's public key")
+		}
+		_, err = ctx.SignatureProvider.HasBeenSignedBy(auditor, verifier)
```

Comment:
If I understand correctly, HasBeenSignedBy has to be called in the right order. Would it make sense to check in any order and keep a counter of how many have been checked already?

Reply:
checking in any order is expensive because we would need to verify the signatures against multiple public keys.
Anyhow, a driver can use a different implementation of `SignatureProvider` as you suggest, sure.